packages/api/src/services/emp-state.ts

```diff
@@ -0,0 +1,132 @@
+import assert from "assert";
+import * as uma from "@uma/sdk";
+import Promise from "bluebird";
+const { Emp } = uma.clients;
+import { BigNumber, utils } from "ethers";
+const { parseBytes32String } = utils;
+import { asyncValues } from "../libs/utils";
+import { Json, Libs } from "..";
+
+type Instance = ReturnType<typeof Emp.connect>;
+export default (config: Json, libs: Libs) => {
+  const { registeredEmps, provider, emps } = libs;
+
+  async function calcGcr(params: { totalTokensOutstanding: string | null; totalPositionCollateral: string | null }) {
+    const { totalTokensOutstanding, totalPositionCollateral } = params;
+    assert(uma.utils.exists(totalTokensOutstanding), "requires total tokens outstanding");
+    assert(uma.utils.exists(totalPositionCollateral), "requires total position collateral");
+    if (BigNumber.from(totalTokensOutstanding).gt(0)) {
+      return utils.parseUnits(totalPositionCollateral).div(totalTokensOutstanding);
```

Comment:
hope this is correct. also is there an example where we calc gcr for tokens/collateral with varying decimals?

Reply:
hhmm. I dont think what you have here will for for all financial contracts. `utils. parseUnits` without a second parameter converts the input number to 1e18. Numerically, this will work fine if the collateral & synthetics are both 1e18 scaled as you'll basically be going collateral*1e18/debt which works fine if they are both 1e18 scaled.

To do this correctly I think you should pass the number of decimals in the collateral/synthetics as a second param into the `parseUnits` method. Note this assumes you are dealing with the _new_ emp where the collateral and synthetics decimals always match. If we want this to work with all old EMPs as well then you will need to do some more involved normalization. see https://github.com/UMAprotocol/protocol/blob/bb4dd0bff53733aad3ced3ce9e78c94a64aa8161/packages/financial-templates-lib/src/clients/FinancialContractClient.js#L318

However, I don't really think that's needed given almost all of these legacy contracts have expired!

Reply:
ok thanks @chrismaree will update